README.md

```diff
@@ -485,22 +485,17 @@ This behavior is also applied to [`collections.defaultdict`](https://docs.python
 ([1, 2, 3], PyTreeSpec({'a': [*, *], 'b': [*]}))
 ```
 
-Note that there are no restrictions on the `dict` to require the keys are comparable (sortable).
-There can be multiple types of keys in the dictionary.
-The keys are sorted in ascending order by `key=lambda k: k` first if capable otherwise fallback to `key=lambda k: (k.__class__.__qualname__, k)`. This handles most cases.
+**`tree_unflatten` does not preserve the same key ordering as the original pytree passed to `tree_flatten`.**
+While unflattening and reconstructing a new pytree via the tree specification, the key ordering of the output dictionaries is also sorted.
+This means flattening a pytree and then unflattening it back will change the key order.
 
 ```python
->>> sorted({1: 2, 1.5: 1}.keys())
-[1, 1.5]
->>> sorted({'a': 3, 1: 2, 1.5: 1}.keys())
-Traceback (most recent call last):
-    ...
-TypeError: '<' not supported between instances of 'int' and 'str'
->>> sorted({'a': 3, 1: 2, 1.5: 1}.keys(), key=lambda k: (k.__class__.__qualname__, k))
-[1.5, 1, 'a']
+>>> leaves, treespec = optree.tree_flatten({'b': [3], 'a': [1, 2]})
+>>> optree.tree_unflatten(treespec, leaves)
+{'a': [1, 2], 'b': [3]}
 ```
 
-If users want to keep the values in the insertion order, they should use [`collections.OrderedDict`](https://docs.python.org/3/library/collections.html#collections.OrderedDict), which will take the order of keys under consideration:
+If users want to always keep the values in the insertion order in `tree_unflatten`, they should use [`collections.OrderedDict`](https://docs.python.org/3/library/collections.html#collections.OrderedDict), which will take the order of keys under consideration:
```

Comment:
The fact that `OrderedDict` and `dict` behave differently is confusing and unnecessary. Why not follow `tf.nest` / `dm-tree` behavior where:

* All dict behave similarly (no difference between `OrderedDict`, `dict`, `Mapping`, subclasses,...)
* Key are preserved for all operations (not just `map` but also `flatten+unflatten`)

Reply:
> The fact that OrderedDict and dict behave differently is confusing and unnecessary. 

> - Key are preserved for all operations (not just map but also unflatten(flatten()))

This is something we can do. If the performance drop is acceptable, we can make this the default behavior.

------

> All dict behave similarly (no difference between `OrderedDict`, `dict`, `Mapping`, subclasses,...)
> Specially: `Mapping`, subclasses

We intentionally do not support this. The custom subclasses from the user can vary. We want users to handle this themselves.

For example, `OrderedDict` is a subclass of `dict` and it's also a `Mapping`, it considers key order in `__eq__` while `dict` does not.

```python
>>> {'a': 1, 'b': 2} == {'b': 2, 'a': 1}
True
>>> OrderedDict([('a', 1), ('b', 2)]) == OrderedDict([('b', 2), ('a', 1)])
True
```

The Python equivalent for `PyODict_Type` in C:

```python
class OrderedDict(dict):
    def __eq__(self, other):
        '''od.__eq__(y) <==> od==y.  Comparison to another OD is order-sensitive
        while comparison to a regular mapping is order-insensitive.

        '''
        if isinstance(other, OrderedDict):
            return dict.__eq__(self, other) and all(map(_eq, self, other))
        return dict.__eq__(self, other)

    ...
```

The behavior mismatch between `dict.__eq__` and `OrderedDict.__eq__` can be resolved by saving the keyview rather than a list of keys as metadata in the tree node.

In OpTree, we have maintained [a pretty list of Python standard collection types](https://github.com/metaopt/optree#built-in-pytree-node-types).

- [`tuple`](https://docs.python.org/3/library/stdtypes.html#tuple)
- [`list`](https://docs.python.org/3/library/stdtypes.html#list)
- [`dict`](https://docs.python.org/3/library/stdtypes.html#dict)
- [`collections.namedtuple`](https://docs.python.org/3/library/collections.html#collections.namedtuple) and its subclasses
- [`collections.OrderedDict`](https://docs.python.org/3/library/collections.html#collections.OrderedDict)
- [`collections.defaultdict`](https://docs.python.org/3/library/collections.html#collections.defaultdict)
- [`collections.deque`](https://docs.python.org/3/library/collections.html#collections.deque)
- [`PyStructSequence`](https://docs.python.org/3/c-api/tuple.html#struct-sequence-objects) types created by C API [`PyStructSequence_NewType`](https://docs.python.org/3/c-api/tuple.html#c.PyStructSequence_NewType)